scipy/_lib/_array_api.py

```diff
@@ -354,3 +360,16 @@ def xp_unsupported_param_msg(param):
 
 def is_complex(x, xp):
     return xp.isdtype(x.dtype, 'complex floating')
+
+def scipy_namespace_for(xp):
+    if is_numpy(xp):
+        import scipy
+        return scipy
+    elif is_cupy(xp):
+        import cupyx  # type: ignore[Import]
+        return cupyx.scipy
+    elif is_jax(xp):
+        import jax
+        return jax.scipy
+    else:
+        return None
```

Comment:
@izaid here's my implementation. We'll see if CI is happy with me using it in `special._support_alternative_backends`. It seems convenient to me to be returning `None` rather than raising a `RuntimeError` when no matching module is found, but let me know if you think otherwise.

Reply:
> @izaid here's my implementation. We'll see if CI is happy with me using it in `special._support_alternative_backends`. It seems convenient to me to be returning `None` rather than raising a `RuntimeError` when no matching module is found, but let me know if you think otherwise.

That looks pretty good! Two questions:

1) Does that include the array_api_compat libraries as well? Will `is_numpy(xp)` return true for either `numpy` or `array_api_compat.numpy`?

2) Actually I think it should raise an error? I believe `array_api_compat.array_namespace` will error if it can't find an appropriate namespace, so we'd expect the same behaviour.

Also, since you are returning, you don't need the `elif`s! Just several `if`s, can be separated by blank lines.

Reply:
Great! I'm of the general opinion that if + return is better than elif, because it then lets us escape the indent and do other stuff if needed.

If you want to return `None`, that's fine. My gut tells me it should raise an error or return the actual SciPy as a default, but I don't have a use case in mind for that.

I think we'll need to add a second argument at some point to help with submodules. This is also easy, but I can do it in a follow-up PR. I'm imagining use like this
```
xp = array_api_compat.array_namespace(x, y, ...)
spx = scipy_namespace_for(xp, import_names = ['sparse', 'special'])

# spx.sparse or spx.special now just works
```

Without that second argument, the submodules won't necessarily have been imported and it would fail on `spx.special`. This is exactly what I've used for a very long time.

Reply:
Ok, I actually do have a use case in mind for other instances. I'd say we should default to returning the actual SciPy. Essentially then we're saying "we don't know if these other packages have a SciPy, but if it is convertible to NumPy it will just work". In this case, only CuPy and Jax need if cases then.

Reply:
> Without that second argument, the submodules won't necessarily have been imported and it would fail on spx.special. This is exactly what I've used for a very long time.

This is actually not true anymore! We can access submodules directly now, see https://scientific-python.org/specs/spec-0001/.

> I'd say we should default to returning the actual SciPy. Essentially then we're saying "we don't know if these other packages have a SciPy, but if it is convertible to NumPy it will just work". In this case, only CuPy and Jax need if cases then.

Okay, that sounds reasonable. How would you write the logic in `special._support_alternative_backends`, then?

Reply:
> > Without that second argument, the submodules won't necessarily have been imported and it would fail on spx.special. This is exactly what I've used for a very long time.
> 
> This is actually not true anymore! We can access submodules directly now, see https://scientific-python.org/specs/spec-0001/.

Oh, that's interesting! Is that standard though or just recommended?

> > I'd say we should default to returning the actual SciPy. Essentially then we're saying "we don't know if these other packages have a SciPy, but if it is convertible to NumPy it will just work". In this case, only CuPy and Jax need if cases then.
> 
> Okay, that sounds reasonable. How would you write the logic in `special._support_alternative_backends`, then?

What function specifically? Happy to take a look!

Reply:
> What function specifically? Happy to take a look!

```python
def get_array_special_func(f_name, xp, n_array_args):
    spx = scipy_namespace_for(xp)
    if is_numpy(xp):
        f = getattr(_ufuncs, f_name, None)
    elif is_torch(xp):
        f = getattr(xp.special, f_name, None)
    elif is_cupy(xp):
        import cupyx  # type: ignore[import]
        f = getattr(cupyx.scipy.special, f_name, None)
    elif xp.__name__ == f"{array_api_compat_prefix}.jax":
        f = getattr(xp.scipy.special, f_name, None)
    elif spx:
        f = getattr(spx.special, f_name, None)
    else:
        f_scipy = getattr(_ufuncs, f_name, None)
        def f(*args, **kwargs):
            array_args = args[:n_array_args]
            other_args = args[n_array_args:]
            array_args = [np.asarray(arg) for arg in array_args]
            out = f_scipy(*array_args, *other_args, **kwargs)
            return xp.asarray(out)
    return f
```

> Oh, that's interesting! Is that standard though or just recommended?

SciPy has adopted it, so you can do

```console
import scipy
scipy.special.{...}
```

Reply:
Oops, I left in the `-` lines from the diff, I've edited it to be correct now.

Reply:
> > What function specifically? Happy to take a look!
> 
> ```python
> def get_array_special_func(f_name, xp, n_array_args):
>     spx = scipy_namespace_for(xp)
>     if is_numpy(xp):
>         f = getattr(_ufuncs, f_name, None)
>     elif is_torch(xp):
>         f = getattr(xp.special, f_name, None)
>     elif is_cupy(xp):
>         import cupyx  # type: ignore[import]
>         f = getattr(cupyx.scipy.special, f_name, None)
>     elif xp.__name__ == f"{array_api_compat_prefix}.jax":
>         f = getattr(xp.scipy.special, f_name, None)
>     elif spx:
>         f = getattr(spx.special, f_name, None)
>     else:
>         f_scipy = getattr(_ufuncs, f_name, None)
>         def f(*args, **kwargs):
>             array_args = args[:n_array_args]
>             other_args = args[n_array_args:]
>             array_args = [np.asarray(arg) for arg in array_args]
>             out = f_scipy(*array_args, *other_args, **kwargs)
>             return xp.asarray(out)
>     return f
> ```
> 
> > Oh, that's interesting! Is that standard though or just recommended?
> 
> SciPy has adopted it, so you can do
> 
> ```
> import scipy
> scipy.special.{...}
> ```

Oh, I see. Is the problem here is that torch puts its 'special' in the main namespace, rather than have a separate 'scipy' one? In that case, I'd basically bite the bullet and do something like `get_special_namespace_for` and also for any other offending namespaces. I wouldn't have something get the funcs by name, should just return the actual namespace object.

Reply:
But for NumPy arrays, we can't just use `scipy.special`, since those functions are now the wrapped versions we are using this in - instead we need to pull from `_ufuncs`. It seems to make more sense to combine them into one.

Also see my comment above, those cases for CuPy and JAX shouldn't be there in the code you quoted.